room ban info endpoint

majestrate opened the pull request:
Fetch room permissions routes. Basic unit tests and basic functionality.

Fixes #63

Commit:
Fetch room perminssions routes. Basic unit tests and basic functionality.

jagerman commented on sogs/model/room.py:
```diff
@@ -1464,8 +1464,34 @@ def unpin(self, msg_id: int, admin: User):
+            ):
+                data = dict()
+                for k in row.keys():
+                    if k not in ('session_id', 'room', 'user'):
```
```suggestion
                    if k not in ('session_id', 'room', 'user'):
                        if k == 'banned' and not row[k]:
                            continue  # only include banned when true
```

majestrate committed:
Update sogs/model/room.py

add suggested case by jason

Co-authored-by: Jason Rhinelander <jason@imaginary.ca>

jagerman commented on sogs/routes/rooms.py:
```diff
@@ -278,6 +278,24 @@ def update_room(room):
+    dict of session_id to current permission info
+
+    """
+    return jsonify(room.export_permissions(g.user))
```
I'm worried this might get big, as is, because it'll be something like:
```json
{"15abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789": {"read": true, "upload": false},"15bcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789a": {"banned": true},
...}
```

It might be worth encoding it to something smaller, perhaps by instead of a dict of "perm": true|false, we could return a set (I dunno if jsonify is happy converting a set to a list; might need to be done manually) containing:
"r"/"-r" - included if read is not null
"w"/"-w" included if write is not null
"a"/"-a" included if accessible is not null
"u"/"-u" included if upload is not null

Some of the flags are non-nullable (because they are additive rather than overriding) and so wouldn't have "-" variations, i.e. would only be included when true:
"b" would be included only if banned is true
"A" would be included only if admin is true
"M" would be included only if moderator is true *and* admin is not true
"V" would be included only if moderator is true and visible_mod is true

jagerman commented on sogs/model/room.py:
```diff
@@ -1464,8 +1464,36 @@ def unpin(self, msg_id: int, admin: User):
+                data = dict()
+                for k in row.keys():
+                    if k not in ('session_id', 'room', 'user'):
+                        if k == 'banned' and not row[k]:
```
My bad, there are actually a few of these non-nullables:
```suggestion
                        if k in ('banned', 'moderator', 'admin', 'visible_mod') and not row[k]:
```

jagerman commented on sogs/model/room.py:
```diff
@@ -1464,8 +1464,36 @@ def unpin(self, msg_id: int, admin: User):
+        if not self.check_moderator(mod):
+            app.logger.warning("unable to get room permissions for user")
+            raise BadPermission()
+        with db.transaction():
```
There's only a single query here, no need for a transaction.

majestrate committed:
Update sogs/model/room.py

Co-authored-by: Jason Rhinelander <jason@imaginary.ca>

Commit:
dont use a transaction

Commit:
* add future permissions
* make permissions and future permissions an attribute
* add unit tests

jagerman commented on sogs/model/room.py:
```diff
@@ -1494,6 +1491,33 @@ def export_permissions(self, mod):
+                if k in ('at', 'session_id'):
+                    data[k] = row[k]
+                else:
+                    data[k] = bool(row[k])
```
This needs to preserve `None`s as `None`

jagerman replied:
Never mind, misread the diff.

Commit:
respect none value

jagerman commented on sogs/routes/rooms.py:
```diff
@@ -278,6 +278,61 @@ def update_room(room):
+    if compact is set to 1 this will be a string,
+    otherwise it will be a dict containing the name of the permission mapped to a boolean value.
+    """
+    transform = lambda x: x  # noqa: E731
+    if request.args.get("compact", type=int, default=0):
+        transform = compact_permissions
+    return jsonify(transform(room.permissions))
```
```suggestion
    perm = room.permissions
    if request.args.get("compact", type=int, default=0):
        perm = compact_permissions(perm)
    return jsonify(perm)
```

Commit:
do not include keys in dict we do not care about like session_id room and user

Commit:
get rid of banal compaction of permissions, we'll use xz if we want compression later

Commit:
add function to fix up permissions in routes codepath

add extra members and delete redundant info from model

Commit:
clarify docstrin

majestrate committed:
dont filter data in model

jagerman committed:
Tweak addExtraPermInfo to remove some stuff we can omit

jagerman committed:
Fix permission ban query

jagerman committed:
Minor postgresql migration fixes

- we could leave "empty" (i.e. all null) rows in the
  user_permission_futures table
- if something goes screwy and the user_permission_futures_pkey doesn't
  exist (i.e. because we already dropped it) then the update was
  failing; added an `IF EXISTS` to avoid the error.  (This is unlikely
  to happen in production, but doesn't hurt).
